src/execution/collectFields.ts

```diff
@@ -34,15 +34,17 @@ export interface DeferUsage {
 export interface FieldDetails {
   node: FieldNode;
   deferUsage: DeferUsage | undefined;
+  fragmentVariableValues?: ObjMap<unknown> | undefined;
 }
 
 interface CollectFieldsContext {
   schema: GraphQLSchema;
   fragments: ObjMap<FragmentDefinitionNode>;
-  variableValues: { [variable: string]: unknown };
   operation: OperationDefinitionNode;
   runtimeType: GraphQLObjectType;
   visitedFragmentNames: Set<string>;
+  localVariableValues: { [variable: string]: unknown } | undefined;
```

Comment:
Is it safe to store the `localVariableValues` in the context? Might this not get rewritten by nested fragments?

Reply:
The goal was to not rewrite it which is successful due to recalculating at every intersection, this also is meant to support using the variables in a fragment-definition in a future spread.

Reply:
What I'm worried about is a document like the following:

```graphql
        query {
          ...A(value: "A")
        }
        fragment A($value: String!) on Query {
          ...B(value: "B")
          someFieldA(input: $value)
        }
        fragment B($value: String!) on Query {
          someFieldB(input: $value)
        }
}

When I coalesced the parameters to CollectFieldsImpl into a `context`, the intent was to do so because these variables were essentially readonly (e.g. schema) or accumulators (e.g. newDeferUsages) and so there could be no conflicts.
```

Reply:
This should all be safe because of how we reset this context during collectFields, every spread entry will either take this to undefined _or_ overwrite it with the defined variables

Reply:
I see, the test passes, I didn't catch/realize that that even though you are resetting the `localVariableValues` on the context for the fragment, the later fields will pick up the value from the `localVariableValues` that were already destructured here: https://github.com/JoviDeCroock/graphql-js/blob/2d354bcd7d567c71cd2c535a9dde25fd5d260ad4/src/execution/collectFields.ts#L159

Because you are using the context to basically pass a different value to the inner `collectFieldsImpl`, I would still suggest refactoring as an additional argument, but you are correct that there is no bug in this case.

Reply:
if you would like here is a PR with those tests to show that they pass: https://github.com/JoviDeCroock/graphql-js/pull/8